internal/pkg/cli/run_local.go

```diff
@@ -455,6 +466,119 @@ func (o *runLocalOpts) getTask(ctx context.Context) (orchestrator.Task, error) {
 	return task, nil
 }
 
+func (o *runLocalOpts) prepareTask(ctx context.Context) (orchestrator.Task, error) {
+	task, err := o.getTask(ctx)
+	if err != nil {
+		return orchestrator.Task{}, fmt.Errorf("get task: %w", err)
+	}
+
+	mft, _, err := workloadManifest(&workloadManifestInput{
+		name:         o.wkldName,
+		appName:      o.appName,
+		envName:      o.envName,
+		ws:           o.ws,
+		interpolator: o.newInterpolator(o.appName, o.envName),
+		unmarshal:    o.unmarshal,
+		sess:         o.envManagerSess,
+	})
+	if err != nil {
+		return orchestrator.Task{}, err
+	}
+
+	containerURIs, err := o.buildContainerImages(mft)
+	if err != nil {
+		return orchestrator.Task{}, fmt.Errorf("build images: %w", err)
+	}
+
+	// replace built images with the local built URI
+	for name, uri := range containerURIs {
+		ctr, ok := task.Containers[name]
+		if !ok {
+			return orchestrator.Task{}, fmt.Errorf("built an image for %q, which doesn't exist in the task", name)
+		}
+
+		ctr.ImageURI = uri
+		task.Containers[name] = ctr
+	}
+
+	return task, nil
+}
+
+func (o *runLocalOpts) watchLocalFiles(stopCh <-chan struct{}) (<-chan interface{}, <-chan error, error) {
+	copilotDir := o.ws.Path()
+
+	watchCh := make(chan interface{})
+	watchErrCh := make(chan error)
+
+	watcher, err := o.newRecursiveWatcher()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err = watcher.Add(copilotDir); err != nil {
+		return nil, nil, err
+	}
+
+	watcherEvents := watcher.Events()
+	watcherErrors := watcher.Errors()
+
+	debounceTimer := time.NewTimer(o.debounceTime)
+	if !debounceTimer.Stop() {
+		// flush the timer in case stop is called after the timer finishes
+		<-debounceTimer.C
+	}
+
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				watcher.Close()
+				return
+			case err, ok := <-watcherErrors:
+				watchErrCh <- err
+				if !ok {
+					watcher.Close()
+					return
+				}
+			case event, ok := <-watcherEvents:
+				if !ok {
+					watcher.Close()
+					return
+				}
+
+				// skip chmod events
+				if event.Has(fsnotify.Chmod) {
+					break
+				}
+
+				// check if any subdirectories within copilot directory are hidden
+				isHidden := false
+				parent := copilotDir
+				suffix, _ := strings.CutPrefix(event.Name, parent+"/")
+				for _, child := range strings.Split(suffix, "/") {
```

Comment:
I have a few questions/concerns -
1. If `evenv.Name` is "a/b/something" - not start with "a/b/copilot/", then is it still worth it to loop through "a", "b, "something", checking "a/b/copilot/a", "a/b/copilot/a/b", "a/b/copilot/a/b/something" front of each?
2. `"/"` does not work across OS. What do you think of using [`filepath.Split`](https://pkg.go.dev/path/filepath#Split)?
3. Can you help me understand the intention behind this block - why do we ignore only the hidden file under `a/b/copilot`?
4. Where do we send an event to `watchCh`, so that the runner restarts the containers?

Reply:
1. I don't think I understand your question here 🤔 
2. The watcher actually only sends updates in the form "/a/b/something" regardless of the OS being used - this is a specification of fsnotify. I used `filepath.Split` at first but because of this it broke in windows.
3. My thought process was that a user is likely running this in the root of their workspace - where it's likely in a git repository. Because of how the watcher works and how we get the ws path, it should be global. The issue I was having was that the watcher would receive file updates from `.git/(something)` and this would cause it to keep rebuilding because each time it built the containers it would cause this file to be changed. My solution was to make it so that all files in non hidden directories/non hidden files themselves in the parent directory will be updated.
4. After this block when the debounce timer is reset, that eventually causes the `watchCh` to receive `nil`.